presto-main/src/main/java/com/facebook/presto/execution/executor/MultilevelSplitQueue.java

```diff
@@ -0,0 +1,168 @@
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+@ThreadSafe
+public class MultilevelSplitQueue
+{
+    public static final int[] LEVELS = {0, 1, 10, 60, 300};
+
+    private final List<PriorityQueue<PrioritizedSplitRunner>> levelWaitingSplits;
+    private final AtomicLong[] levelMinPriority;
+
+    private final ReentrantLock lock;
+    private final Condition notEmpty;
+
+    private final List<CounterStat> selectedLevelCounters;
+
+    private final AtomicInteger size = new AtomicInteger(0);
+
+    public MultilevelSplitQueue()
+    {
+        this.levelMinPriority = new AtomicLong[LEVELS.length];
+        this.levelWaitingSplits = new ArrayList<>(LEVELS.length);
+        ImmutableList.Builder<CounterStat> counters = ImmutableList.builder();
+
+        for (int i = 0; i < LEVELS.length; i++) {
+            levelMinPriority[i] = new AtomicLong(-1);
+            levelWaitingSplits.add(new PriorityQueue<>());
+            counters.add(new CounterStat());
+        }
+
+        this.selectedLevelCounters = counters.build();
+        this.lock = new ReentrantLock();
+        this.notEmpty = lock.newCondition();
+    }
+
+    public void offer(PrioritizedSplitRunner split)
+    {
+        checkArgument(split != null, "split is null");
+
+        split.setReady();
+        lock.lock();
+
+        levelWaitingSplits.get(split.getLevel()).offer(split);
+        size.incrementAndGet();
+        notEmpty.signal();
+
+        lock.unlock();
+    }
+
+    public PrioritizedSplitRunner take()
+            throws InterruptedException
+    {
+        while (true) {
+            lock.lockInterruptibly();
+            try {
+                PrioritizedSplitRunner result;
+                while ((result = pollFirstSplit()) == null) {
+                    notEmpty.await();
+                }
+
+                if (result.updateLevelPriority()) {
+                    offer(result);
+                    continue;
+                }
+
+                int selectedLevel = result.getLevel();
+                levelMinPriority[selectedLevel].set(result.getLevelPriority());
+                selectedLevelCounters.get(result.getLevel()).update(1);
+                size.decrementAndGet();
+
+                return result;
+            }
+            finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    @GuardedBy("lock")
+    private PrioritizedSplitRunner pollFirstSplit()
+    {
+        for (PriorityQueue<PrioritizedSplitRunner> level : levelWaitingSplits) {
+            PrioritizedSplitRunner split = level.poll();
+            if (split != null) {
+                return split;
+            }
+        }
+
+        return null;
+    }
+
+    public void remove(PrioritizedSplitRunner split)
+    {
+        checkArgument(split != null, "split is null");
+        lock.lock();
+        for (PriorityQueue<PrioritizedSplitRunner> level : levelWaitingSplits) {
+            level.remove(split);
```

Comment:
Since you manually maintain `size` now, shouldn't you decrement it (conditionally) here?

(I'm not a bug fan of maintaining `size` btw, i liked earlier approach better)